tools/pubsys/src/aws/client.rs

```diff
@@ -1,159 +1,165 @@
-use async_trait::async_trait;
+use aws_config::profile::ProfileFileCredentialsProvider;
+
+use aws_config::{ConfigLoader, SdkConfig};
+use aws_sdk_ec2::{Endpoint, Region};
+use aws_sdk_sts::model::Credentials as StsCredentials;
+use aws_sdk_sts::{Client as StsClient, Credentials as SdkCredentials};
 use pubsys_config::AwsConfig;
-use rusoto_core::{request::DispatchSignedRequest, HttpClient, Region};
-use rusoto_credential::{
-    AutoRefreshingProvider, AwsCredentials, CredentialsError, DefaultCredentialsProvider,
-    ProfileProvider, ProvideAwsCredentials,
-};
-use rusoto_ebs::EbsClient;
-use rusoto_ec2::Ec2Client;
-use rusoto_ssm::SsmClient;
-use rusoto_sts::{StsAssumeRoleSessionCredentialsProvider, StsClient};
-use snafu::ResultExt;
-
-pub(crate) trait NewWith {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static;
-}
+use snafu::{OptionExt, ResultExt};
+use std::convert::TryInto;
 
-impl NewWith for EbsClient {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+const PROVIDER_NAME: &str = "pubsys";
 
-impl NewWith for Ec2Client {
-    fn new_with<P, D>(request_dispatcher: D, credentials_provider: P, region: Region) -> Self
-    where
-        P: ProvideAwsCredentials + Send + Sync + 'static,
-        D: DispatchSignedRequest + Send + Sync + 'static,
-    {
-        Self::new_with(request_dispatcher, credentials_provider, region)
-    }
-}
+pub(crate) async fn build_client_config<S1, S2>(
+    region: S1,
+    sts_region: S2,
+    aws: &AwsConfig,
```

Comment:
This should have a more descriptive name:
```suggestion
    pubsys_aws_config: &PubsysAwsConfig,
```